runtime/src/main/java/org/corfudb/runtime/collections/StreamListenerFullSync.java

```diff
@@ -0,0 +1,81 @@
+package org.corfudb.runtime.collections;
+
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.CorfuStoreMetadata.Timestamp;
+import org.corfudb.runtime.exceptions.StreamingException;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.exceptions.UnreachableClusterException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.IntervalRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import java.util.List;
+
+/**
+ * This is an abstract stream listener callback implementation for clients interested in automatic
+ * re-subscription upon any error. The re-subscription policy of this implementation protects against data
+ * loss by immediately performing a full sync on any type of error.
+ *
+ * Created by annym on 06-22-2021
+ */
+@Slf4j
+public abstract class StreamListenerFullSync implements StreamListener {
+
+    private final CorfuStore store;
+    private final String namespace;
+    private final String streamTag;
+    private final List<String> tablesOfInterest;
+
+    public StreamListenerFullSync(CorfuStore store, String namespace, String streamTag, List<String> tablesOfInterest) {
+        this.store = store;
+        this.namespace = namespace;
+        this.streamTag = streamTag;
+        this.tablesOfInterest = tablesOfInterest;
+    }
+
+    public StreamListenerFullSync(CorfuStore store, String namespace, String streamTag) {
+        this(store, namespace, streamTag, null);
+    }
+
+    public void subscribe() {
+
+        Timestamp fullSyncTimestamp = performFullSync();
+
+        if (tablesOfInterest == null) {
+            store.subscribeListener(this, namespace, streamTag, fullSyncTimestamp);
+        } else {
+            store.subscribeListener(this, namespace, streamTag, tablesOfInterest, fullSyncTimestamp);
+        }
+    }
+
+    protected abstract Timestamp performFullSync();
```

Comment:
Would be great to have a comment on this method about how they should implement it.